src/models/contact/index.js

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright 2020 Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import {
+  createSolidDataset,
+  getSolidDataset,
+  getSourceUrl,
+  getThingAll,
+  getUrl,
+  getUrlAll,
+  saveSolidDatasetAt,
+  setThing,
+  setUrl,
+} from "@inrupt/solid-client";
+import { foaf, rdf, schema, vcard } from "rdf-namespaces";
+import { vcardExtras } from "../../addressBook";
+import { joinPath } from "../../stringHelpers";
+import { ERROR_CODES, isHTTPError } from "../../error";
+
+/**
+ * Contacts represent the dataset in a user's AddressBook, e.g. /contacts/Person/<unique-id>/index.ttl#this
+ *
+ * @typedef Contact
+ * @type {object}
+ * @property {object} dataset - The dataset that the thing lives in
+ * @property {object} thing - The contact itself, be that a person, a group, or something else
+ */
+
+/* Model constants */
+export const NAME_EMAIL_INDEX_PREDICATE = vcardExtras("nameEmailIndex");
+export const NAME_GROUP_INDEX_PREDICATE = vcardExtras("groupIndex");
+export const INDEX_FILE = "index.ttl";
+export const PEOPLE_INDEX_FILE = "people.ttl";
```

Comment:
GROUP constants should live in models/group, and PEOPLE constants should live in a models/personContact (which doesn't exist.) (While you're at it, rename models/group to models/groupContact for clarity.)

I would suggest removing the type_map and indexes, and create `saveGroupInAddressBook()` in models/group and `savePersonInAddressBook()` methods in a new `models/personContact`, which call `addContactIndexToAddressBook` with the appropriate type. That way, `contact` doesn't know or care about types, the actual typed contacts do.